src/main/java/com/gtnewhorizon/gtnhlib/capability/item/AbstractInventorySourceIterator.java

```diff
@@ -0,0 +1,146 @@
+package com.gtnewhorizon.gtnhlib.capability.item;
+
+import java.util.Objects;
+
+import net.minecraft.init.Items;
+import net.minecraft.inventory.IInventory;
+import net.minecraft.item.ItemStack;
+
+/**
+ * A InventorySourceIterator implementation for something IInventory-like.
+ */
+public abstract class AbstractInventorySourceIterator implements InventorySourceIterator {
+
+    private final int[] slots;
+
+    private int i = 0, last = 0;
+    private boolean dirty = false;
+
+    private final FastImmutableItemStack pooled = new FastImmutableItemStack(null);
+
+    protected AbstractInventorySourceIterator(int[] slots) {
+        this.slots = slots;
+    }
+
+    /**
+     * @see IInventory#getStackInSlot(int)
+     */
+    protected abstract ItemStack getStackInSlot(int slot);
+
+    /**
+     * @see IInventory#setInventorySlotContents(int, ItemStack)
+     */
+    protected abstract void setInventorySlotContents(int slot, ItemStack stack);
+
+    /**
+     * Marks the backing inventory dirty. Only called the first time a stack is set.
+     * @see IInventory#markDirty()
+     */
+    protected void markDirty() {
+
+    }
+
+    @Override
+    public boolean hasNext() {
+        return i < slots.length;
+    }
+
+    @Override
+    public ImmutableItemStack next() {
+        last = i++;
+
+        pooled.stack = getStackInSlot(slots[last]);
+        return pooled.stack == null ? null : pooled;
+    }
+
+    @Override
+    public boolean hasPrevious() {
+        return i > 0;
+    }
+
+    @Override
+    public ImmutableItemStack previous() {
+        last = --i;
+
+        pooled.stack = getStackInSlot(slots[last]);
+        return pooled.stack == null ? null : pooled;
+    }
+
+    @Override
+    public int nextIndex() {
+        return i + 1;
+    }
+
+    @Override
+    public int previousIndex() {
+        return i - 1;
+    }
+
+    @Override
+    public ItemStack extract(int amount) {
+        ItemStack inSlot = getStackInSlot(slots[last]);
+
+        if (!isStackValid(inSlot)) return null;
+
+        int toExtract = Math.min(inSlot.stackSize, amount);
+
+        ItemStack extracted = inSlot.splitStack(toExtract);
+
+        setInventorySlotContents(slots[last], inSlot.stackSize == 0 ? null : inSlot);
+
+        if (!dirty) {
+            markDirty();
+            dirty = true;
+        }
+
+        return extracted;
+    }
+
+    @Override
+    public void insert(ItemStack stack) {
+        if (!isStackValid(stack)) return;
+
+        ItemStack inSlot = getStackInSlot(slots[last]);
```

Comment:
I'm very confused by that `last` thing. Can't we just have `current` instead of two indices?